plugins/ssh/src/main/kotlin/io/paddle/plugin/ssh/SshCommandExecutor.kt

```diff
@@ -0,0 +1,87 @@
+package io.paddle.plugin.ssh
+
+import com.github.fracpete.processoutput4j.output.StreamingProcessOutput
+import com.github.fracpete.rsync4j.RSync
+import com.github.fracpete.rsync4j.Ssh
+import io.paddle.execution.CommandExecutor
+import io.paddle.plugin.ssh.output.RemoteOutputOwner
+import io.paddle.project.Project
+import io.paddle.terminal.Terminal
+import io.paddle.terminal.TextOutput
+import io.paddle.utils.ext.Extendable
+import java.io.File
+
+class SshCommandExecutor(private val host: String, private val user: String,
+                         private var remoteDir: String, output: TextOutput) : CommandExecutor(OutputConfiguration(output)) {
+    object Extension : Project.Extension<SshCommandExecutor> {
+        override val key: Extendable.Key<SshCommandExecutor> = Extendable.Key()
+
+        override fun create(project: Project): SshCommandExecutor {
+            val host: String? = project.config.get("executor.host")
+            val user: String? = project.config.get("executor.user")
+            val dir: String? = project.config.get("executor.directory")
+            return SshCommandExecutor(host!!, user!!, dir!!, project.output)
+        }
+    }
+
+    init {
+        if (!remoteDir.endsWith("/")) {
```

Comment:
Instead of making remoteDir `var `, I would recommend that you declare it as `val` and get in constructor just argument. Then you can write `val remoteDir = if (remoteDir.endsWith("/")) remoteDir + '/' else remoteDir`

Reply:
Good point, fixed.